generator/osm2meta.cpp

```diff
@@ -472,6 +472,182 @@ std::string MetadataTagProcessorImpl::ValidateAndFormat_duration(std::string con
   return format(hours);
 }
 
+std::string MetadataTagProcessorImpl::MaxChargingPower(std::string_view const & m, std::string const & k, std::string const & v)
+{
+  bool typedc = false;
+  char split = ' ';
+  size_t pos = 0, last = 0, mult = 0, watts = 0;
+  float numeric = 0.0, maximum = 0.0;
+  std::string dc, ac, lower = v, part, result;
+
+  std::string value = static_cast<std::string>(m);
+  if (v.find("http") != std::string::npos)
+    return value;
+  if (MAX_POWER_POINT != '.')
+    std::replace(dc.begin(), dc.end(), MAX_POWER_POINT, '.');
+
+  if (value.size() > 1) {
+    if ((pos = value.find(MAX_POWER_DELIM)) != std::string::npos) {
+      dc = value.substr(0, pos - (sizeof(MAX_POWER_DCSTR) - 1));
+      ac = value.substr(pos + sizeof(MAX_POWER_DELIM) - 1, value.size() - (pos + sizeof(MAX_POWER_DELIM) + sizeof(MAX_POWER_ACSTR) - 2));
+    } else if ((pos = value.find(MAX_POWER_DCSTR)) != std::string::npos) {
+      dc = value.substr(0, pos);
+    } else if ((pos = value.find(MAX_POWER_ACSTR)) != std::string::npos) {
+      ac = value.substr(0, pos);
+    }
+  }
+
+  while ((pos = lower.find("к")) != std::string::npos) {
+    lower[pos] = 'k';
+    lower[pos + 1] = ' ';
+  }
+
+  while ((pos = lower.find("К")) != std::string::npos) {
+    lower[pos] = 'k';
+    lower[pos + 1] = ' ';
+  }
+
+  while ((pos = lower.find("в")) != std::string::npos) {
+    lower[pos] = 'w';
+    lower[pos + 1] = ' ';
+  }
+
+  while ((pos = lower.find("В")) != std::string::npos) {
+    lower[pos] = 'w';
+    lower[pos + 1] = ' ';
+  }
+
+  std::transform(lower.begin(), lower.end(), lower.begin(), ::tolower);
+  lower.erase(std::remove_if(lower.begin(), lower.end(), ::isspace), lower.end());
+
+  if ((pos = lower.find(';')) != std::string::npos) {
+    split = ';';
+  } else if ((pos = lower.find('+')) != std::string::npos) {
+    split = '+';
+  } else if ((pos = lower.find('-')) != std::string::npos) {
+    split = '-';
+  } else if ((pos = lower.find('=')) != std::string::npos) {
+    split = '=';
+  } else if ((pos = lower.find(':')) != std::string::npos) {
+    split = ':';
+  } else if ((pos = lower.find('/')) != std::string::npos) {
+    split = '/';
+  } else if ((pos = lower.find('\\')) != std::string::npos) {
+    split = '\\';
+  } else if ((pos = lower.find(',')) != std::string::npos) {
+    if (((mult = lower.find("kw")) != std::string::npos) && (lower.find("kw", mult + 1) != std::string::npos)) {
+      split = ',';
+    } else {
+      split = ' ';
+      pos = lower.size();
+    }
+  } else {
+    pos = lower.size();
+  }
+
+  if ((k.find("combo") != std::string::npos) || (k.find("chademo") != std::string::npos) ||
+      (k.find("super") != std::string::npos) || (k.find("bosch") != std::string::npos) ||
+      (k.find("gb_dc") != std::string::npos) || (k.find("3105") != std::string::npos) ||
+      (k.find("roadster") != std::string::npos))
+    typedc = true;
+
+  do {
```

Comment:
This loop looks fragile to parse any random data coming from OpenStreetMap.org 

What is the main logic/idea behind it? Can it be simplified?